pkg/storegateway/bucket.go

```diff
@@ -262,10 +262,28 @@ func NewBucketStore(
 
 // RemoveBlocksAndClose remove all blocks from local disk and releases all resources associated with the BucketStore.
 func (s *BucketStore) RemoveBlocksAndClose() error {
+	return s.removeBlocksAndClose(context.Background(), false)
+}
+
+// RemoveBlocksCloseAndWait remove all blocks from local disk, releases all resources associated with the BucketStore
+// and waits until all dependencies have been stopped.
+func (s *BucketStore) RemoveBlocksCloseAndWait(ctx context.Context) error {
+	return s.removeBlocksAndClose(ctx, true)
+}
```

Comment:
we should be having this synchronous in all cases. If the snapshotter shuts down async, then it can create a file after the tenant directory [has been cleaned up](https://github.com/grafana/mimir/blob/a6fa9b031a9e0419cf783bf489738f1b19399ff3/pkg/storegateway/bucket_stores.go#L577-L588). Previously `RemoveBlocksAndClose` has been synchronous anyways.

having said that, it's also ok to not necessarily propagate the context all the way to `RemoveBlocksCloseAndWait`. This will be done soon with #8389

Reply:
> having said that, it's also ok to not necessarily propagate the context all the way to RemoveBlocksCloseAndWait.

I think we need to, actually. At least for a case when the first `syncBlocks` fails, the `Snapshotter` won't start. In this case, the call to `Snapshotter.Stop` will block, waiting on an empty channel.

With the above at hand, I wonder if it would be simpler to replace `Snapshotter.stopped` channel with a WaitGroup:

```go
func Start() {
  wg.Add(1)
  go func() {
      defer wg.Done()
   }
}

func Stop() {
  close(stop)
  wg.Wait()
}
```

Although, the above it a (nit), given the plans for https://github.com/grafana/mimir/issues/8389

[1]: https://github.com/grafana/mimir/blob/865d1651f7ddde4f841929fb74d3b506bacf055e/pkg/storegateway/bucket.go#L348-L350

Reply:
ok, let's keep the current implementation. This is a nightmare to navigate